applications: serial_lte_modem: improve PPP documentation (#14313)
nrfconnect/sdk-nrf: applications/serial_lte_modem/doc/PPP_AT_commands.rst

  #XPPP: 0,0

  OK

Testing on Linux
================

> You can test SLM's PPP on Linux by using the ``pppd`` command.

Reviewer: I think that this would benefit on having the overlay's which you have to use in here as well, or linked to here.

Reviewer: Also, do I need to run some AT-commands for CMUX or PPP before doing this? If yes, this should give an example of those commands. I got this to work previously, but cannot remember anymore how, and the documentation leaves this open for interpretation.

  #XPPP: 0,0

  OK

Testing on Linux
================

You can test SLM's PPP on Linux by using the ``pppd`` command.

> Having PPP running on SLM and its UART connected to the Linux host, run the following command:

Reviewer 2: ```suggestion
With PPP running on SLM and its UART connected to the Linux host, run the following command:
```

   .. code-block:: console

      $ sudo pppd -detach <PPP_UART_dev> <baud_rate> noauth crtscts novj nodeflate nobsdcomp debug +ipv6 usepeerdns noipdefault defaultroute defaultroute6 ipv6cp-restart 5 ipcp-restart 5

> Make sure to replace ``<PPP_UART_dev>`` by the device file assigned to the PPP UART (e.g. :file:`/dev/ttyACM1`) and ``<baud_rate>`` by the baud rate of the UART that PPP is using.

Reviewer 2: ```suggestion
Make sure to replace ``<PPP_UART_dev>`` by the device file assigned to the PPP UART (for example, :file:`/dev/ttyACM1`) and ``<baud_rate>`` by the baud rate of the UART that PPP is using.
```

> After running the command and the PPP link negotiation completes successfully, a new network interface will be created (typically ``ppp0``); it will send and receive IP traffic through the modem of the nRF91 Series SiP running SLM.

Reviewer 2: ```suggestion
After you have run the command and the PPP link negotiation has completed successfully, a new network interface will be created (typically ``ppp0``).
This network interface will send and receive IP traffic through the modem of the nRF91 Series SiP running SLM.
```

.. note::

>    You may encounter some issues with DNS resolution.

Reviewer 2: ```suggestion
   You can encounter some issues with DNS resolution.
```

>    They can be worked around by manually editing :file:`/etc/resolv.conf`.

Reviewer 2: ```suggestion
   Edit the :file:`/etc/resolv.conf` file to work around these issues.
```

>    You can put in DNS servers that are reachable with your current network configuration.

Reviewer 2: ```suggestion
   You can add DNS servers that are reachable with your current network configuration.
```

> 
>    Those can even be the DNS servers that SLM's PPP sends as part of the PPP link negotiation.
>    They are the DNS servers of the default PDN connection obtained from the modem.

Reviewer 2: ```suggestion
   These added servers can even be the DNS servers that SLM's PPP sends as part of the PPP link negotiation, which are the DNS servers of the default PDN connection obtained from the modem.
```
Note the removed empty line so we don't end up with a one-sentence paragraph.

Author: Is it forbidden? :smiley: (Curious to know the actual rationale.)

Reviewer 2: Not forbidden, but it's generally not good form in writing. In the same way as we try to avoid really long-running sentences, but sometimes they are required (like here I actually made one by combining two shorter sentences).

We still use one-sentence paragraphs as introductions to procedure steps and such, and sometimes for emphasis or visibility.

  #XPPP: 0,0

  OK

Testing on Linux
================

You can test SLM's PPP on Linux by using the ``pppd`` command.

For that, SLM must be compiled without CMUX because there is no widely available utility that allows multiplexing a device file on Linux.
If you happen to be equipped with one, you can use SLM's PPP with ``pppd`` through CMUX by first setting up CMUX and replacing the device file in the ``pppd`` command by that of SLM's PPP channel (that will have been created by the CMUX utility).

> Assuming that SLM's UARTs are connected to the Linux host, you first need to get PPP running.
> For that, start SLM and issue an ``AT+CFUN=1`` command.

Reviewer 2: ```suggestion
You first need to get PPP running.
For that, start SLM and issue an ``AT+CFUN=1`` command.
```
I moved the first part higher.

Can you format these as numbered steps, preferably so that a step isn't just "Run the following command", but rather either "Run the following command to [...]" or "[Do something]. To do this, run the following command." so the reader knows what they are doing when they run the commands.

Make sure that the network registration succeeds and that PPP is started successfully, either by having a look at SLM's logs or issue an ``AT#XPPP?`` command that should return ``#XPPP: 1,0``.

> Then, run the following command:

Author: I mean, are you gonna run a `sudo` command on SLM? :smile:  (Still added your precision.)

   .. code-block:: console

      $ sudo pppd -detach <PPP_UART_dev> <baud_rate> noauth crtscts novj nodeflate nobsdcomp debug +ipv6 usepeerdns noipdefault defaultroute defaultroute6 ipv6cp-restart 5 ipcp-restart 5

> Make sure to replace ``<PPP_UART_dev>`` by the device file assigned to the PPP UART (for example, :file:`/dev/ttyACM1`) and ``<baud_rate>`` by the baud rate of the UART that PPP is using.

Reviewer: And possibly that with 9161 with only PPP, the default PPP_UART_dev is /dev/ttyACM1 and with 9160 it is /dev/ttyACM2.

> Make sure to replace ``<PPP_UART_dev>`` by the device file assigned to the PPP UART (for example, :file:`/dev/ttyACM1`) and ``<baud_rate>`` by the baud rate of the UART that PPP is using.

Author: > Could mention that without CMUX the baud rate is 1000000.

I am still looking into the baud rates. Here I would rather point the user to the configuration than state that the baud rate is X even though the default `overlay-ppp-without-cmux.overlay` sets it to 1000000.
Though to address your comment I added this information to the overlay's description.

> After you have run the command and the PPP link negotiation has completed successfully, a new network interface will be created (typically ``ppp0``)
> This network interface will send and receive IP traffic through the modem of the nRF91 Series SiP running SLM.

Reviewer 2: ```suggestion
After you have run the command and the PPP link negotiation has completed successfully, a new network interface will be created, typically ``ppp0``.
This network interface will send and receive IP traffic through the modem of the nRF91 Series SiP running SLM.

If you have an utility that allows multiplexing a device file on Linux, you can use SLM's PPP with the ``pppd`` command through CMUX.
To do this, you must set up CMUX, and then replace the device file in the ``pppd`` command with SLM's PPP channel, which will have been created by the CMUX utility.
```

Author: The CMUX part here is coming too late to me because if CMUX is used then it must be set up *before* PPP, so before running the `pppd` command.

Reviewer 2: This could be something for a note before the steps, then. That way it's clear that it's something that can be done but modifies the noted steps.

Reviewer 2: My thinking, though not clearly communicated, was that the part about CMUX would not be part of numbered steps, but rather content after them.

Author: Yeah, I was also thinking about making it a note before the numbered steps start.

Author: By the way, let me correct you for once: `a` utility, not `an`. :wink:

  #XPPP: 0,0

  OK

Testing on Linux
================

You can test SLM's PPP on Linux by using the ``pppd`` command, though SLM must be compiled without CMUX because there is no widely available utility that allows multiplexing a device file on Linux.

.. note::

   If you have a utility that allows multiplexing a device file on Linux, you can use SLM's PPP with the ``pppd`` command through CMUX.
   To do this, you must first set up the CMUX link.
   Then, make sure to replace the device file argument in the ``pppd`` command with that of SLM's PPP channel, which will have been created by the CMUX utility.
   See :ref:`SLM_AT_CMUX` for more information on SLM's CMUX.

For the process described here, SLM's UARTs must be connected to the Linux host.

1. Get PPP running on SLM.
   For that, start SLM and issue an ``AT+CFUN=1`` command.

>    Make sure that the network registration succeeds and that PPP is started successfully, either by having a look at SLM's logs or issuing an ``AT#XPPP?`` command that should return ``#XPPP: 1,0``.

Author: This step is part of getting PPP running on SLM. Do you still think it should be a separate step?

>    Make sure that the network registration succeeds and that PPP is started successfully, either by having a look at SLM's logs or issuing an ``AT#XPPP?`` command that should return ``#XPPP: 1,0``.

Reviewer 2: Technically, PPP can start without checking that it did, so I feel it makes sense as a separate step.

>    Make sure that the network registration succeeds and that PPP is started successfully, either by having a look at SLM's logs or issuing an ``AT#XPPP?`` command that should return ``#XPPP: 1,0``.

Author: Indeed.

   To do this, you must first set up the CMUX link.
   Then, make sure to replace the device file argument in the ``pppd`` command with that of SLM's PPP channel, which will have been created by the CMUX utility.
   See :ref:`SLM_AT_CMUX` for more information on SLM's CMUX.

For the process described here, SLM's UARTs must be connected to the Linux host.

1. Get PPP running on SLM.
   To do this, start SLM and issue an ``AT+CFUN=1`` command.

#. Make sure that the network registration succeeds and that PPP is started successfully.
   To do this, either look at SLM's logs, or issue an ``AT#XPPP?`` command, which returns ``#XPPP: 1,0`` when PPP has started successfully.

#. Run the following command on the Linux host:

   .. code-block:: console

      $ sudo pppd -detach <PPP_UART_dev> <baud_rate> noauth crtscts novj nodeflate nobsdcomp debug +ipv6 usepeerdns noipdefault defaultroute defaultroute6 ipv6cp-restart 5 ipcp-restart 5

   Replace ``<PPP_UART_dev>`` by the device file assigned to the PPP UART and ``<baud_rate>`` by the baud rate of the UART that PPP is using.

>    Typically, when ``uart1`` is assigned to be the PPP UART (in the devicetree overlay), the device file assigned to it is :file:`/dev/ttyACM2` for an nRF9160 DK, and :file:`/dev/ttyACM2` for the other nRF91 Series DKs.

Reviewer: Second /dev/ttyACM2 -> /dev/ttyACM1.

Also, how about mentioning the typical 1000 000 baud rate in here? Maybe better in here than in the configurations.

This is already merged, so fix at least the first one in the PR that is open.

Author: Well spotted.
